docs/develop-an-agent.md

```diff
@@ -0,0 +1,423 @@
+
+#### **Step 1** Get identty information
+
+Send the request to the identity endpoint `/run/aziot/identityd.sock`
+
+If the connection is refused, retry after few seconds
+
+_Prerequisite_: You may need to first install `jq` (e.g. `sudo apt install jq`)
+
+```bash
+# Calling IS as the root user to skip the need to add my user to the requisite security groups for IS/KS/CS. 
+identity="$(sudo curl --unix-socket '/run/aziot/identityd.sock' 'http://identityd.sock/identities/modules/mymodule?api-version=2020-09-01&type=aziot')" 
+
+# From the response retrieve the hubname, gatewayHost, device and module identity, authentication type, and the key handle
+hub_name="$(<<< "$identity" jq -r '.spec.hubName')"
+gateway_host="$(<<< "$identity" jq -r '.spec.gatewayHost')"
+device_id="$(<<< "$identity" jq -r '.spec.deviceId')"
+module_id="$(<<< "$identity" jq -r '.spec.moduleId // ""')"
+auth_type="$(<<< "$identity" jq -r '.spec.auth.type')"
+key_handle="$(<<< "$identity" jq -r '.spec.auth.keyHandle')"
+```
+
+#### **Step 2** Build a bas64 encoded resource uri
+
+When generating the signed token to connect to IoT Hub, the **message** parameter given to the Key Service's `/sign` API is a base-64 encoded value. It is created by concatenating the `resource_uri` with your desired token expiration. For example:
+
+```bash
+
+# URL encode the resource of hub + device + module
+# If we were doing this for the device identity the resource_uri wouldn't include the '/modules/%s'
+resource_uri="$(printf '%s/devices/%s/modules/%s' "$hub_name" "$device_id" "$module_id" | jq -sRr @uri)"
+
+# Using an expiry of 1 day from today
+expiry="$(bc <<< "$(date +%s) + 60 * 60 * 24")"
+
+# Construct the full message to be signed which is a url encoded resource URI concatenated with the expiry and then base64 encoded
+signature_data="$(printf '%s\n%s' "$resource_uri" "$expiry" | base64 -w 0)"
+```
+
+More details on how the IoT Hub service interprets this signed message can be found in the docs on [Security tokens](https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-security#security-tokens). 
+
+
+#### **Step 3** Get a signed token from the keys service
+
+Send a request to the key service endpoint `/run/aziot/keyd.sock`.
+
+If the connection is refused, go back to Step 1 since IoT Hub device information could have changed.
+
+Otherwise, parse the `signature` from the resposne.
+
+```bash
+# Request the key service sign the message and then parse the 'signature' from the response and url decode it
+signature="$(
+    sudo curl --unix-socket '/run/aziot/keyd.sock' \
+        -X POST -H 'content-type: application/json' --data-raw "$(
+            jq -cn --arg 'keyHandle' "$key_handle" --arg 'message' "$signature_data" '{
+                "keyHandle": $keyHandle,
+                "algorithm": "HMAC-SHA256",
+                "parameters": {
+                    "message": $message,
+                },
+            }'
+        )" 'http://keyd.sock/sign?api-version=2020-09-01' |
```

Comment:
We recently bumped the API version for KS. `2020-09-01` will still work, but I advise updating this to the latest.

Reply:
AFAIK we haven't released a patch or updated package for 1.2 that would include the new API.  Until then I'll plan to use the minimum version that works.  

In our API docs we should be vigilant about calling out which APIs are available on older versions and not just the latest.  We may not have a way to easily gate updates to the docs with our releases, but it's easier for someone consuming the docs if the docs are in sync with what is officially released rather than the current state of what is checked in.